aspnet/fundamentals/app-state.rst

```diff
@@ -1,8 +1,258 @@
+
+For example, some simple :doc:`middleware` could add something to the ``Items`` collection:
+
+.. code-block:: c#
+
+	app.Use(async (context, next) =>
+		{
+			// perform some verification
+			context.Items["isVerified"] = true;
+			await next.Invoke();
+		});
+
+and later in the pipeline, another piece of middleware could access it:
+
+.. code-block:: c#
+
+	app.Run(async (context) =>
+	{
+		await context.Response.WriteAsync("Verified request? "
+			+ context.Items["isVerified"]);
+	});
+
+.. note:: Since keys into ``Items`` are simple strings, if you are developing middleware that needs to work across many applications, you may wish to prefix your keys with a unique identifier to avoid key collisions (e.g. "MyComponent.isVerified" instead of just "isVerified").
+
+Installing and Configuring Session
+----------------------------------
+
+ASP.NET 5 ships a session package that provides middleware for managing session state. You can install it by including a reference to the package in your project.json file:
+
+.. literalinclude:: app-state/sample/src/AppState/project.json
+	:language: javascript
+	:linenos:
+	:lines: 5-14
+	:emphasize-lines: 5-6
+
+Once the package is installed, Session must be configured in your application's ``Startup`` class. Session is built on top of ``IDistributedCache``, so you must configure this as well, otherwise you will receive an error.
+
+.. note:: If you do not configure at least one ``IDistributedCache`` implementation, you will get an exception stating "Unable to resolve service for type 'Microsoft.Framework.Caching.Distributed.IDistributedCache' while attempting to activate 'Microsoft.AspNet.Session.DistributedSessionStore'."
+
+ASP.NET ships with several implementations of ``IDistributedCache``, including an in-memory option (to be used during development and testing only). To configure session using this in-memory option, add the following to ``ConfigureServices``:
+
+.. code-block:: c#
+
+	services.AddCaching();
+	services.AddSession();
+
+Then, add the following to ``Configure`` and you're ready to use session in your application code:
+
+.. code-block:: c#
+
+	app.UseSession();
+
+You can reference Session from ``HttpContext`` once it is installed and configured.
+
+.. note:: If you attempt to access ``Session`` before ``UseSession`` has been called, you will get an ``InvalidOperationException`` exception stating that "Session has not been configured for this application or request."
+
+.. warning:: If you attempt to access ``Session`` after you have already begun writing to the ``Response`` stream, you will get an ``InvalidOperationException`` as well, stating that "The session cannot be established after the response has started". This exception may not be displayed in the browser; you may need to view the web server log  to discover it, as shown below:
```

Comment:
This only happens for new sessions. E.g. if the session cookie has not yet been.